cluster/src/main/java/io/atomix/cluster/impl/DefaultClusterService.java

```diff
@@ -104,12 +110,19 @@ public Node getLocalNode() {
 
   @Override
   public Set<Node> getNodes() {
-    return ImmutableSet.copyOf(nodes.values());
+    return ImmutableSet.copyOf(nodes.values()
+        .stream()
+        .filter(node -> node.type() == Node.Type.DATA || node.getState() == State.ACTIVE)
+        .collect(Collectors.toList()));
   }
 
   @Override
   public Node getNode(NodeId nodeId) {
-    return nodes.get(nodeId);
+    Node node = nodes.get(nodeId);
+    if (node != null && node.type() == Node.Type.CLIENT && node.getState() == State.INACTIVE) {
```

Comment:
What about data type here?

Reply:
If correct the method name should be clearer.

Reply:
This logic is just filtering tombstones, which shouldn't be visible via the API. It's an implementation detail of the replication algorithm.